theano/tensor/nlinalg.py

```diff
@@ -742,14 +742,18 @@ def __init__(self, ind=2):
     def make_node(self, a):
         a = as_tensor_variable(a)
         out_dtype = a.dtype
-        out = theano.tensor.matrix(dtype=out_dtype)
+        out = theano.tensor.tensor4(dtype=out_dtype)
```

Comment:
okay, thanks for the hint, now it is working with a.type(), and I added tests for other tensors, infer_shape seems to works.